File: backend/voting-app/src/main/java/com/tu/votingapp/utils/mappers/RoleMapper.java
package com.tu.votingapp.utils.mappers;

import com.tu.votingapp.dto.general.RoleDTO;
import com.tu.votingapp.entities.RoleEntity;
import org.mapstruct.Mapper;

@Mapper(componentModel = "spring")
public interface RoleMapper {
    RoleDTO toDto(RoleEntity roleEntity);

    RoleEntity toEntity(RoleDTO roleDTO);
}

File: frontend/android/app/src/main/java/com/example/android/utils/mappers/RoleMapper.kt
package com.example.android.utils.mappers

import com.example.android.dto.response.RoleDTO
import com.example.android.entity.RoleEntity
import com.example.android.utils.mappers.PermissionMapper.toDto
import com.example.android.utils.mappers.PermissionMapper.toEntity

object RoleMapper {

    fun RoleDTO.toEntity(): RoleEntity = RoleEntity(
        id = this.id,
        name = this.name
    ).also { role ->
        role.permissions = this.permissions.map { it.toEntity() }
    }

    fun RoleEntity.toDto(): RoleDTO = RoleDTO(
        id = this.id!!,
        name = this.name,
        permissions = permissions.map { i -> i.toDto() }
    )
}